labs/Demos/RailFence.md

```diff
@@ -0,0 +1,133 @@
+```cryptol
+module labs::Demos::RailFence where
+
+/** constraint on number of additional rails (excluding first rail) */
+type constraint _r_ r = (fin r, r >= 1)
+/** constraint on message length */
+type constraint _m_ m = (fin m, m >= 1)
+
+/** width of cycle indices */
+type cw r = max 2 (width (2*r - 1))
+/** width of message indices */
+type pw m r = width (max m (max (2*r) (m /^ (2*r) * (2*r))))
+
+/** Swap `i`th and `j`th entries of sequence `a` */
+swap : {n, i, a} (fin i, i == width n + 1) => [n]a -> [i] -> [i] -> [n]a
+swap a i j = updates a [i,j] (a @@ [j,i])
+
+/** indices per cycle */
+cycle: {r} _r_ r => ([cw r], [r-1][2][cw r], [cw r])
+cycle = (0, transpose [take`{r-1} ([1...]:[inf][cw r]), reverse (take`{r-1} ([`(r+1)...]:[inf][cw r]))], `r)
+
+/** inverse a 1:1 mapping of permuted values [0..`(n-1)] */
```

Comment:
“inverse a”?

Reply:
Nouning verbs is all the rage. I still need to write a narrative around this.